Remove urn_state type

paytonrules opened the pull request:
I did this as three migrations. The first to replace the all_urns function with a much simpler one. The second to remove all references to urn_state, and the third to rename obsolete_urn_state to obsolete_urn_snapshot. The code changes (non SQL edition) are almost all around tests.

paytonrules committed:
Replace all_urns with a simpler query using snapshots

This replaces the complicated all_urns with a much simpler SQL call that
goes right at the urn_snapshots, now that we have them.

paytonrules committed:
Remove references to urn_state and the type

In addition remove the all_urn_states query test because that query
goes away, the get_urn_query_test because that query also goes away, and
the computed columns tests because those functions go away as well. Wee!

Finally convert any place that uses urn_state and use urn_snapshot
directly in the return types.

paytonrules committed:
Rename obsolete_urn_state to obsolete_urn_snapshot

rmulhol commented on db/migrations/20200504135835_remove_urn_state.sql:
```diff
@@ -0,0 +1,226 @@
+DROP FUNCTION IF EXISTS api.get_urn;
+
+DROP FUNCTION IF EXISTS api.bite_event_urn;
+CREATE OR REPLACE FUNCTION api.bite_event_urn(event api.bite_event) RETURNS api.urn_snapshot
```
Do we need the `DROP FUNCTION` if we're doing `CREATE OR REPLACE`?

paytonrules replied:
Unfortunately since we change the signature of the function, we have to drop it first. You can't change the return value with a replace.

We probably don't need `CREATE OR REPLACE` - I was being overly-cautious..

rmulhol commented on transformers/component_tests/queries/get_urn_query_test.go:
```diff
@@ -1,157 +0,0 @@
-		diffID                 int64
-	)
-
-	const getUrnQuery = `SELECT urn_identifier, ilk_identifier, ink, art, created, updated FROM api.get_urn($1, $2, $3)`
```
I like the idea of using the `api.urn_snapshots` table to replace this query, but do you know how we'd structure the query against that table to guarantee we're getting the most recent one by block number in GraphQL?

rmulhol replied:
Worst come to worst I could see a case for keeping the query and it just wraps the `order by block_number desc limit 1`

paytonrules replied:
I guess it's a question if people are using it - we could certainly use the query you just said to keep it around.

paytonrules replied:
I restored get_urn - assuming all_urns only returns the latest unique urns (ergo distinct on urn id and ilk) then get_urn should always return only 1 urn. 

I'd like to have the equivalent of ASSERT somewhere here - since if it returns more than one then something is wrong in the all_urns query. In either case as long as get_urn uses all_urns then we shouldn't need to `order by block_number desc limit`.

paytonrules replied:
(Assuming I got it right of course).

rmulhol commented on transformers/component_tests/queries/urn_state_computed_columns_test.go:
```diff
@@ -1,283 +0,0 @@
-	. "github.com/onsi/gomega"
-)
-
-var _ = Describe("Urn state computed columns", func() {
```
Just confirming - these go away because we have the exact same computed columns on the `urn_snapshot` table?

paytonrules replied:
Yup 👍

rmulhol commented on transformers/component_tests/queries/all_urn_states_query_test.go:
```diff
@@ -1,218 +0,0 @@
-	. "github.com/onsi/gomega"
-)
-
-var _ = Describe("Urn history query", func() {
```
And here it goes 💥 because we can just `select * from api.urn_snapshot where urn_identifier = x and ilk_identifier = y`?

paytonrules replied:
Exactly.

rmulhol commented on db/migrations/20200504135835_remove_urn_state.sql:
```diff
@@ -0,0 +1,226 @@
+SELECT * FROM api.urn_snapshot
+WHERE ilk_identifier = (SELECT identifier FROM maker.ilks WHERE ilks.id = flip.ilk_id)
+AND urn_identifier = (SELECT identifier FROM maker.urns WHERE urns.id = flip.urn_id)
+AND block_height <= flip.block_height
```
I think this may need to `order by block_height desc limit 1` now that we're querying from a table that could have multiple rows with matching `urn_identifier`/`ilk_identifier`. Probably applies to queries below as well

paytonrules replied:
I restored get)urn - which should take care of that.

rmulhol reviewed (commented):
Looking really good! Planning to take a closer look later, but dropping some initial thoughts

rmulhol commented on db/migrations/20200504131504_update_all_urns_function.sql:
```diff
@@ -0,0 +1,80 @@
+CREATE OR REPLACE FUNCTION api.all_urns(block_height bigint DEFAULT api.max_block(), max_results integer DEFAULT NULL::integer, result_offset integer DEFAULT 0) RETURNS SETOF api.urn_snapshot
+    LANGUAGE sql STABLE
+    AS $$
+SELECT urn_identifier, ilk_identifier, all_urns.block_height, ink, coalesce(art, 0), created, updated
```
I think we need this to be distinct on the urn_identifier + ilk_identifier so that we're only getting the most recent state of every urn in the system

paytonrules replied:
~Does his mean you can have the same urn_identifier with a different ilk, and that should be a different urn?~

Snapshots, right, duh.

paytonrules replied:
👍

paytonrules committed:
Update timestamps for migrations

paytonrules committed:
Add annotations to urn test helpers

This is so you can tell what field is actually failing.

paytonrules committed:
Make api.all_urns distinct on urn and ilk

To make sure we return the urns in a predictable order I need to make
the query for distinct urn_snapshots a subquery, because the order by
updated only applies within the distinct grouping, to ensure we get the
latest urn. I also added a test for this query.

paytonrules committed:
Restore `get_urn` helper function

Assuming all_urns only returns unique individual urns (as it should)
using this in the other queries should mean you'll always only get one urn.

elizabethengelman commented on transformers/component_tests/queries/all_urns_query_test.go:
```diff
@@ -115,6 +115,39 @@ var _ = Describe("Urn view", func() {
 		helper.AssertUrn(result[1], expectedUrnTwo)
 	})
 
+	It("returns only the most recent urn for multiple snapshots of the same urn", func() {
```
👏

elizabethengelman commented on transformers/component_tests/queries/test_helpers/test_helpers.go:
```diff
@@ -286,13 +286,13 @@ type UrnState struct {
-	Expect(actual.Art).To(Equal(expected.Art))
-	Expect(actual.Created).To(Equal(expected.Created))
-	Expect(actual.Updated).To(Equal(expected.Updated))
+	Expect(actual.UrnIdentifier).To(Equal(expected.UrnIdentifier), "Urn Identifier")
```
🎉

elizabethengelman reviewed (approved):
LGTM, nice work!

I'm a little hesitant to merge this in while we're still working on urn_snapshot reconciliation, but I'd be curious to hear what others think!

rmulhol commented on db/migrations/20200512130930_update_all_urns_function.sql:
```diff
@@ -0,0 +1,83 @@
+                              all_urns.block_height, ink, coalesce(art, 0), created, updated
+           FROM api.urn_snapshot
+           WHERE block_height <= all_urns.block_height
+           ORDER BY urn_identifier, ilk_identifier, updated DESC) AS latest_urns
```
Can we just `ORDER BY updated` here (omitting `urn_identifier, ilk_identifier`)? Also, do we need `AS latest_urns`?

paytonrules replied:
Unfortunately we can't omit `urn_identifier` and `ilk_identifier` from ORDER BY because they are in the DISTINCT above. I'm not sure about latest_urns, I have to double-check.

rmulhol commented on db/migrations/20200512141017_remove_urn_state.sql:
```diff
@@ -0,0 +1,232 @@
+    AS $_$
+
+SELECT *
+    FROM api.all_urns(get_urn.block_height)
```
Seems kinda heavyweight to query `all_urns` for one urn - might it be better to:
```
SELECT * FROM api.urn_snapshot WHERE ilk_identifier = x AND urn_identifier = y AND block_height <= n ORDER BY block_height DESC LIMIT 1
```

paytonrules replied:
I was trying to be consistent with the other ones that use all_urns, but you're probably right.

rmulhol commented on db/migrations/20200512141017_remove_urn_state.sql:
```diff
@@ -0,0 +1,232 @@
+       AS $$
+SELECT *
+FROM api.get_urn(
+     (SELECT identifier FROM maker.ilks WHERE ilks.id = flip.ilk_id),
```
Hm, this is _way_ out of scope for this PR, but it seems like `api.get_flip` should be returning a `maker.flip` and we should be deleting `api.flip_bid_snapshot` - basically a story to do exactly what you're doing here for a different place where we've got a type used in queries + a trigger table

rmulhol replied:
Mostly just caught my eye because we have an `ilk_id` and `urn_id` on that type, instead of the identifiers - which seems odd. If we're going to have a custom type for queries, I would have thought we would have done the join to get the identifiers in the query and not exposed internal DB FKs

paytonrules replied:
I'm not quite sure I followed that last bit - is there an action item here besides a story about flip_bid_snapshot?

rmulhol replied:
No, sorry - kinda just thinking out loud. Will create stories for this stuff

rmulhol reviewed (approved):
Looks good to me - a couple minor comments, would be curious to hear your thoughts, nothing merge blocking

paytonrules committed:
get_urn can query urn_snapshots directly
